Restrict constants in patterns

nikomatsakis opened the pull request:
Feature-gate the use of constants in patterns unless those constants have simple types, like integers, booleans, and characters. The semantics of constants in general were never widely discussed and the compiler's current implementation is not broadly agreed upon (though it has many proponents). The intention of adding a feature-gate is to give us time to discuss and settle on the desired semantics in an "affirmative" way.

Because the compiler currently accepts a larger set of constants, this is a backwards incompatible change. This is justified as part of the ["underspecified language semantics" clause of RFC 1122][ls]. A [crater run] found 14 regressions on crates.io, which suggests that the impact of this change on real code would be minimal.

Note: this was also discussed on an [internals thread]. Major points from that thread are summarized either inline or in alternatives.

[Rendered view.](https://github.com/nikomatsakis/rfcs/blob/restrict-constants-in-patterns/text/0000-restrict-constants-in-patterns.md)

[ls]: https://github.com/rust-lang/rfcs/blob/master/text/1122-language-semver.md#underspecified-language-semantics
[crater run]: https://gist.github.com/nikomatsakis/26096ec2a2df3c1fb224
[internals thread]: https://internals.rust-lang.org/t/how-to-handle-pattern-matching-on-constants/2846)

nagisa commented on text/0000-restrict-constants-in-patterns.md:
```diff
@@ -0,0 +1,568 @@
+  | u8 | u16 | u32 | u64 | usize  // unsigned integers
+  | char                          // characters 
+  | bool                          // booleans
+  | (B, ..., B)                   // tuples of builtin types
```
Might be useful to consider including `&str` and arrays as well, since they have a well defined “primitive” memory-representation equality.

nagisa commented on text/0000-restrict-constants-in-patterns.md:
```diff
@@ -0,0 +1,568 @@
+
+Here, although it may well be that `T::X == T::Y`, we can't know for
+sure. So, for consistency, we may wish to treat all constants opaquely
+regardless of whether we are in a generic context or not.
```
This is a pretty good point. If we opt to do the checking for regular constants, the contents of constant becomes a public API – changing these becomes a breaking change, which is pretty much counters almost whole consts’ purpose.

oli-obk replied:
Alternatively, in the future, we could forbid this kind of matching if it cannot be guaranteed that `T::X != T::Y` (which could be generically guaranteed by allowing such a statement in where clauses)

nagisa commented:
Pretty strongly in favour.

retep998 commented:
This RFC is a good idea. Since it only affects matching on constants of non-primitive types, I am in favor of this RFC.

petrochenkov commented on text/0000-restrict-constants-in-patterns.md:
```diff
@@ -0,0 +1,568 @@
+-- the same issues arise if you consider exhaustiveness checking.
+
+On the other hand, it feels very silly for the compiler not to
+understand that `match some_bool { true => ..., false => ... }` is
```
I'd say `true` and `false` feel more like unit variants of "`enum bool { false, true }`" than opaque constants.

mahkoh commented:
This breaks lots of code of this kind:

```rust
pub struct Errno(pub c_int);

pub const Interrupted: Errno = Errno(EINTR);
```

My code contains at least 26 instances of this pattern (excluding flags) and an uncountable number of variants:

```
ack "pub struct .*\(pub .*\)" | grep -iv flags | wc -l
```

For instance, the Errno struct comes with 132 variants. How am I supposed to repair this?

petrochenkov commented:
>pub struct Errno(pub c_int);

Newtypes like this are exactly the pattern which is responsible for the most of breakage on crates.io and in rustc itself.

mahkoh commented:
I've also no idea what this has to do with constants.

```rust
match x {
    Errno(1) => 1,
    _ => 0,
}

if x == Errno(1) { 1 } else { 0 }
```

are not guaranteed to produce the same result either. Whether `Errno(1)` is behind a constant or not is irrelevant here.

mahkoh commented:
Of course the same also applies to enums. E.g.

```rust
enum A {
    X(u8),
    Y,
}

impl PartialEq for A {
    fn partial_cmp(&self, other: &A) -> bool { false }
}
```

Matching will return results that are incompatible with the PartialEq results.

mahkoh commented:
Given that the behavior for enums is already fixed (at least the RFC doesn't suggest otherwise), and given that newtypes with all fields public are very similar to enums, it seems that the current behavior (which agrees with the enum behavior) is already the expected behavior.

nagisa commented:
@mahkoh the problem here is using enums has a well specified structural equivalency, but using a constant doesn’t make the equivalency used obvious. There’s a bunch of additional reasons outlined by the RFC why structural equivalency with constants is not satisfactory.

nikomatsakis commented:
@mahkoh 

> I've also no idea what this has to do with constants.

I see a difference between matching on a constant `C = Errno(1)` and matching on `Errno(1)` directly in the pattern. Basically, to me, the pattern translates into a set of "test-and-extract" operations that will be performed. So, the pattern `Errno(pattern>)` compiles to "test that the variant is `Errno` and extract the value". We then recursively apply `<pattern>` to that value. Sometimes there are no values to extract: if you match a nullary enum variant like `None`, it just means "test that the variant is None".  Along these lines, matching a constant pattern like `C` (resp. `1`) corresponds to "test that the value is equal (for some definition of equal) to `C` (resp. 1) and extract no values". 

Looking at it like this, there is no reason to expect that

```rust
match foo { Errno(1) => ... }
```

and 

```rust
const C: Errno = Errno(1);
match foo { C => ... }
```

would do the same thing, just as you would not necessarily expect that:

```rust
if errno.0 == 1 { ... }
```

and 

```rust
if errno == C { ... }
```

would do the same thing.

PS, I am summarizing something I [wrote on the internals thread](https://internals.rust-lang.org/t/how-to-handle-pattern-matching-on-constants/2846/48?u=nikomatsakis), just for reference.

nikomatsakis commented:
@mahkoh 

> This breaks lots of code of this kind

Yes, the proposal is backwards incompatible, and this is exactly the kind of cases that would no longer work. You would have to either use the feature-gate or translate such code from `match foo { ERRFOO => ... }` to `match foo { c if c == ERRFOO => ... }`.

mahkoh commented:
@nikomatsakis 

I'm fine with a feature gate if it can be enabled at the point where the type is defined. E.g.

```rust
#[structural_match]
pub struct Errno(pub c_int);
```

nikomatsakis commented:
@mahkoh 

> I'm fine with a feature gate if it can be enabled at the point where the type is defined.

By this, do you mean that the clients who are matching on `Errno` values would not need to use the feature-gate?

mahkoh commented:
@nikomatsakis Exactly.

nikomatsakis commented:
@mahkoh Hmm. It is normally something we would not allow, since if we decided to use semantic equality (for example), then there might not be an equivalent behavior to `#[structural_match]` in the future (and I wouldn't want to preserve the attribute). OTOH, since I suspect you do define `Eq` for errno to act exactly as structural match acts, that might be ok (modulo exhaustiveness). I have to think about it.

mahkoh commented:
@nikomatsakis As long as the attribute is behind a feature gate, breaking it again doesn't seem to be a problem.

seanmonstar commented:
This sort of breaking change seems like it should be a semver major bump. It's not a security, bug, or soundness fix. It's instead trying to improve reasoning about code. Not that the goal is bad, but that breakage for that goal seems unacceptable for the 1.x versions.

mahkoh commented:
@nikomatsakis Whether the behavior is the current one or one that uses PartialEq, the code I'm worried about behaves the same. This attribute is simply supposed to bridge the gap.

nikomatsakis commented:
@seanmonstar 

> This sort of breaking change seems like it should be a semver major bump. It's not a security, bug, or soundness fix. It's instead trying to improve reasoning about code. Not that the goal is bad, but that breakage for that goal seems unacceptable for the 1.x versions.

Personally, I consider this a bug fix. That is, I did not expect that constants of arbitrary types should be matchable. In fact, I [opened a bug about it before 1.0](https://github.com/rust-lang/rust/issues/20489), but that bug was accidentally closed when I wrote a comment like "does not try to fix #20489", and hence it dropped off of my radar when triaging for "things that ought to be feature-gated before 1.0". However, clearly there is room for disagreement here.

petrochenkov commented:
Curiously, `#[structural_match]` already exists and is called `#[derive(PartialEq)]`.
`#[derive(PartialEq)]` is essentially an assertion that structural and semantic matching do the same thing, so there's no choice left what to do when performing a `match` and we can unambiguously generate the same code as today.
Maybe `#[derive(PartialEq)]` should not only generate some code during expansion, but also emit a flag `this_adt_is_usable_in_const_pattern_matching`,which is considered in later stages of compilation.
`this_adt_is_usable_in_const_pattern_matching` can become a separate attribute eventually if necessary, but generating it as a part of `#[derive(PartialEq)]` allows to avoid breakage and generally looks like a reasonable way forward, covering most of practical cases.

Exhaustiveness checking is a separate orthogonal concern, it still may be better to turn it off for constants regardless of `this_adt_is_usable_in_const_pattern_matching`

nikomatsakis commented:
@petrochenkov 

> Curiously, #[structural_match] already exists and is called #[derive(PartialEq)].

That is mostly true, but only if all types embedded within the struct also `#[derive(PartialEq)]`.

petrochenkov commented:
@nikomatsakis 
`#[structural_match]` should not affect nested things, otherwise you could make an "opaque" structure transparent/matchable by using a newtype:
```
`#[structural_match]`
struct Transparent(Opaque);
```
So, yes, it was implied, that the both `#[structural_match]` or `#[derive(PartialEq)]` have to be checked recursively.
In theory marking a structure containing non-`#[structural_match]` fields with `#[structural_match]` can be made an error, but `derive` is not smart enough to emit `#[structural_match]` conditionally to avoid this error, so I don't consider this variant.

nikomatsakis commented:
@petrochenkov @mahkoh 

So, on reflection, I quite like this idea. My assumption is that this would be something like `#[fundamental]` -- that is, an attribute that we intend to never stabilize but which lets us adopt the subset of semantics we know we want while we dicker about the remainder. In this case, using the semantics you proposed, we could ultimately resolve this in at least two ways:

1. Removing the attribute and adopt the current "structural match" implementation.
2. Removing the attribute and adopt the "semantic equality" interpretation. (*)

Shall I adjust the RFC? I think so.

(*) Actually, the attribute could even stay as a kind of performance optimization. That is, @pnkfelix and I have talked about the compiler recognizing when the `PartialEq` impl resulted from derive and generating better match code in that case, since its semantics are well understood. One could imagine adopting an unsafe attribute or something along those lines, which `PartialEq` implicitly provides.

petrochenkov commented:
`#[structural_match]` can probably be a (possibly unsafe) structural trait (OIBIT) and not an attribute.

petrochenkov commented:
On interaction of `#[structural_match]` with floats.
(I assume that `match` can be used only with `#[structural_match]` types)

Variant 1.
The attribute gives hard guarantees about `PartialEq` properties.
In this case it can be used for optimizations, for example operators `==` for `#[structural_match]` types without padding can be translated into `memcmp`s.
In this case floats simply can't be used in match, their equality comparison is not trivial.

Vartiant 2.
The attribute is a rubber stamp, it says "yeah, this can be used in match" and that's all. `match` in its turn performs a structural comparison OR runs `partial_eq` even if they do different things. In this case floats are allowed in `match` and compared structurally OR semantically, respectively.

Variant 3.
Floats are an exception, they are not `#[structural_match]`, but permitted in `match` and compared structurally or semantically. Structures containing floats are not permitted in match,

--------------------------------

Some bikeshedding, names for `structural_match`:
`trivially_comparable`, `default_comparable` - "comparable" can be confused with `PartialOrd`
`trivial_equality`
`trivial_partial_eq` (or `TrivialPartialEq` for a trait) - says exactly what it does, seems like a good if it provides hard guarantees.

nikomatsakis commented:
Ah, one complication that I just thought of while editing the RFC: this is not just about what types can be used in pattern arms. It is also about exhaustiveness checking. While it seems unlikely that this will have a big impact in practice, the current approach of structural matching also informs exhaustiveness, so we would still be leaving room for regressions if we chose to adopt semantic equality. We could fix this by restricting the exhaustiveness check -- I will have to investigate how that can best be done (but it seems achievable).

eddyb commented:
I missed this RFC being created, but here are my thoughts:
The initial implementation of MIR building resulted in the creation of `PartialEq::eq` calls for matching against a constant, which is dangerous in the presence of interior mutability, and I was against doing that unless there was some stronger guarantee (maybe `unsafe trait MatchEq: Eq {}`).

That said, other than the backwards compatibility issues, I agree that this RFC (i.e. restricting structural equality to known well-behaved types) is a good thing.

There seems to be some discussion around `#[structural_match]` for the opt-in...
Why not types with public contents, recursively? Then the desugaring of constants by inlining is always possible in the presence of privacy.

I like the `#[derive(PartialEq)]` idea, but shouldn't it be `Eq` instead?
I'm nota also sure whether to require public contents if we do that (maybe only for observing exhaustiveness?).

nikomatsakis commented:
Regarding partialeq vs eq, I don't think choosing eq really helps in any
particular respect. If we want to continue supporting floating point
constants (which is debatable) then partial eq is required.

The reason eq doesn't help is that it is not an unsafe trait so we cannot
rely on it to TRULY be eq in any case.

OTOH it might justify some match optimizations (if your Eq impl is broken,
match may panic, just as sort or binary tree will fail). Seems sketchy
though.

I'm fine choosing eq just to be that much more conservative. Another reason
to do so is that the "detect man" heuristic we use today would break for
assoc constants.

I will try to revise rfc asap. The major point of uncertainty for me is how
to handle exhaustiveness. Will write more on this when not on a phone
though :)
On Jan 12, 2016 12:49 PM, "Eduard-Mihai Burtescu" <notifications@github.com>
wrote:

> I missed this RFC being created, but here are my thoughts:
> The initial implementation of MIR building resulted in the creation of
> PartialEq::eq calls for matching against a constant, which is dangerous
> in the presence of interior mutability, and I was against doing that unless
> there was some stronger guarantee (maybe unsafe trait MatchEq: Eq {}).
>
> That said, other than the backwards compatibility issues, I agree that
> this RFC (i.e. restricting structural equality to known well-behaved types)
> is a good thing.
>
> There seems to be some discussion around #[structural_match] for the
> opt-in...
> Why not types with public contents, recursively? Then the desugaring of
> constants by inlining is always possible in the presence of privacy.
>
> I like the #[derive(PartialEq)] idea, but shouldn't it be Eq instead?
> I'm nota also sure whether to require public contents if we do that (maybe
> only for observing exhaustiveness?).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/1445#issuecomment-170990570>.
>

nikomatsakis commented:
OK, I took a stab on the T ride home at describing the `#[structural_match]` solution. Per @eddyb's suggestion, I tied this to the `Eq` trait. I don't actually know what the impact would be here, but I expect it would be minimal -- I guess the impact of my previous change was small enough that one could manually inspect the affected crates.

I also left the matter of exhaustiveness checking etc as an unresolved question. My preference would be to be conservative around user-defined types here, but I'm not really clear on how best to implement that.

nagisa commented:
I dislike the proposal to only consider `PartialEq` equality, because that leaves us with no means whatsoever check for structural equality.

I dislike the idea of making a `PartialEq` matching a thing hidden behind the attribute, since it can be invoked simply by using a `p if p == CONSTANT` or something of the sort. This makes the cost of doing comparisons pretty clear (rather than hidden behind some attribute compiler inserts under some specific circumstances), obvious if you will.

nikomatsakis commented:
@nagisa 

> I dislike the proposal to only consider PartialEq equality, because that leaves us with no means whatsoever check for structural equality.

To be clear, the intention of this RFC is not to settle this debate between structural vs semantic equality, but to make room for us to settle it at our leisure.

nikomatsakis commented:
@nagisa That said, while I agree that semantic equality leaves us with no means to check for structural equality, I consider that a good thing, for the reasons enumerated in the RFC -- I'd be interested to hear if you have strong counterarguments to those, but I think I'd be most interested to hear them on a different thread. ;)

nikomatsakis commented:
I'm going to nominate this for consideration for FCP in next week's @rust-lang/lang meeting, since there hasn't been much feedback.

petrochenkov commented on text/0000-restrict-constants-in-patterns.md:
```diff
@@ -0,0 +1,608 @@
+
+### The structural match attribute
+
+We will introduce an attribute `#[structural_match]` which can be
```
The detailed design section doesn't say what primitive types are `#[structural_match]` (and therefore permitted in constants in patterns).

petrochenkov replied:
Ok, I've made a table:

Primitive type  | Is `#[structural_match]`?
------------- | -------------
`bool`  | Yes
`char`  | Yes
`iN`/`uN`  | Yes
`fN`  | No
`Box`  | No
`str`  | Yes (but not `&str`!)
`[T; N]` / `[T; 0]`  | Depends on `T` / Yes
`(T,)` / `()`  | Depends on `T` / Yes
`[T]`  | Depends on `T`
`*constT /*mut T`  | Yes
`&T/&mut T`  | No
`fn` pointers and items  | Yes
trait objects  | No
closures  | No
structs / enums | Specified explicitly, No by default

nikomatsakis replied:
Thanks @petrochenkov. That table looks about right -- I'm not sure about
`&T` though (and `&mut T`). Those could be supported, I think? (They are
today,and they implement `Eq`, right?)

On Fri, Jan 22, 2016 at 1:28 PM, Vadim Petrochenkov <
notifications@github.com> wrote:

> In text/0000-restrict-constants-in-patterns.md
> <https://github.com/rust-lang/rfcs/pull/1445#discussion_r50572017>:
>
> > +will check the actual variant of the value being matched (regardless
> > +of whether `Modes` implements `PartialEq`, which it does not here). On
> > +the other hand, if matching against `C` were to require a `PartialEq`
> > +impl, then it would be illegal. Therefore matching against an *enum
> > +variant* is distinct from matching against a *constant*.
> > +
> > +# Detailed design
> > +[design]: #detailed-design
> > +
> > +The goal of this RFC is not to decide between semantic and structural
> > +equality. Rather, the goal is to restrict pattern matching to that subset
> > +of types where the two variants behave roughly the same.
> > +
> > +### The structural match attribute
> > +
> > +We will introduce an attribute `#[structural_match]` which can be
>
> Ok, I've made a table:
> Primitive type Is #[structural_match]?
> bool Yes
> char Yes
> iN/uN Yes
> fN No
> Box No
> str Yes (but not &str!)
> [T; N] / [T; 0] Depends on T / Yes
> (T,) / () Depends on T / Yes
> [T] Depends on T
> *constT /*mut T Yes
> &T/&mut T No
> fn pointers and items Yes
> trait objects No
> closures No
> structs / enums Specified explicitly, No by default
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/1445/files#r50572017>.
>

petrochenkov replied:
I aimed for strict structural equality, i.e. `==` is equivalent to `memcmp` (modulo padding).
It doesn't hold for references, two references containing different bit patterns can compare equal, just like e.g. floats.

nikomatsakis replied:
@petrochenkov sadly, it also doesn't hold for structs, tuples, etc: consider padding bytes.

petrochenkov replied:
>consider padding bytes

But... I've mentioned them in the message just above!

The benefit of structural equality is the ability to lower `==` into *set* of `memcmp`s on padding-less parts of a type, not into a single `memcmp`.
Actually, I have no idea how useful can this property be in practice.

petrochenkov replied:
On the other hand, can two values differing only in their padding bits be considered structurally equal? That's pretty philosophical question.

oli-obk replied:
Since we might end up putting enum discriminants into padding bits as an optimization, the padding bits should be ignored.

nikomatsakis replied:
@petrochenkov 

> On the other hand, can two values differing only in their padding bits be considered structurally equal? That's pretty philosophical question.

Yes. ;)

nikomatsakis replied:
@petrochenkov 

> But... I've mentioned them in the message just above!

Sorry, guess I didn't read closely enough.

> The benefit of structural equality is the ability to lower == into set of memcmps on padding-less parts of a type, not into a single memcmp.

In any case, this remains true in many cases, but in others you'd have to add a deref and THEN do some memcmps. Seems fine.

petrochenkov replied:
>add a deref and THEN do some memcmps. Seems fine.

I think I agree.
Padding damages the idea with `memcmp` so much, that references don't make much difference after that.
Without padding, the only thing that has to be known about type `T` to generate `==` is `sizeof(T)`.
With padding you have to know all the types which `T` is composed of, and with that information all the references, boxes and floats can be taken into account as well.

aturon commented:
The lang team is moving this RFC into its weeklong **final comment period**. :8ball:

nikomatsakis commented:
So I implemented the changes proposed here -- though I didn't quite get the stability rules correct yet -- and did a crater run. Unfortunately, I've been encountering some reliability issues with crater that led to a relatively large number of "inconclusive" results (I reran the tests once and encountered similar problems; currently attempting a third run). So the following compatibility results are somewhat provisional. The TL;DR is that I know of **six** crates that would break under this proposal, but there 39 inconclusive results. Of those six, I haven't yet examined them closely enough to say what is needed to fix them: probably just adding `#[derive(Eq)]` in a few places.

I did encounter at least one case where the only available fix was to rewrite the match pattern. Specifically, the `Span` type in the compiler (which stores source location information) overrides `PartialEq` and `Eq` to ignore the `expn_info` field (which stores the macro expansion backtrace). I don't know the code well enough to know if this is just an *optimization* or if this is an important semantic detail. Regardless, it means that match patterns like `DUMMY_SP` and `COMMAND_LINE_SP` are not possible, so I rewrote them to use `if` with an `==` check.

My personal take is that this is a net win -- the old code was basically suspicious and probably a latent bug, if not an actual one. That is, what I **suspect** holds in practice is that the `expn_info` fields are always `None` in the cases that matter, so it's not important whether we ignore them or not.  But if this is **not** the case, then there are two possible scenarios:

- The field **should be** ignored: the old code was buggy.
- The field **should not be** ignored: the old code was correct, but very subtle and non-obvious. It would be better in that case to add something like `eq_including_exp_info()` which does a complete equality check, so that it's obvious that this was intentional.

Anyway, this is the [full crater report]. It reports 45 regressions. I examined those manually and found that 39 of them appear to be inconclusive (timeouts accessing GitHub). I found 6 actual regressions:

- [xhtmlchardet](https://tools.taskcluster.net/task-inspector/#eNUQZ5huReK8VPSJ49huiw/0)
- [elfloader](https://tools.taskcluster.net/task-inspector/#T88DLqqOQaufWYhPDOCOIQ/0)
- [elfloader32](https://tools.taskcluster.net/task-inspector/#GQbClz5uRgC1q7Lvyee1jQ/0)
- [pnet](https://tools.taskcluster.net/task-inspector/#hKNptDYFQP65jlX5Whqw5Q/0)
- [snytex_syntax](https://tools.taskcluster.net/task-inspector/#tPsg57LIQ8GAzZHChwzybA/0) (because of the `Span` type discussed above)
- [chrono](https://tools.taskcluster.net/task-inspector/#0IM-5_hpQM2kB4qycUbz7A/0)

[full crater report]: https://gist.github.com/nikomatsakis/52ace71416d73c304a40

nikomatsakis commented:
**Huzzah!** The lang team [has decided](https://docs.google.com/spreadsheets/d/18FtPnEI8RTPdFhGDtj21ITndakEtxVrs8pVzRNx-jL8/edit#gid=0) to accept this RFC.

Some notes from our discussion:

- It would be better to have the full crater results (subject to the usual caveats in any case). However, it seems safe to characterize the impact of this change as "quite small", and there will be a warning cycle in any case. We do know that the previous crater run -- which tested a more stringent rule -- found 14 regressions, so that serves as a kind of imprecise upper bound.
- We decided that it made sense to settle the "unresolved question" concerning exhaustiveness checking by saying that exhaustiveness checking should be **conservative** with respect to user-defined types, but precise with respect to builtin types. This is consistent with the treatment of the "builtin" constants like `true`, `false`, or `22`, but also forwards compatible with adopting either semantic or structural equality in the future.
    - in particular, if we do NOT handle exhaustiveness, then the RFC isn't really achieving its "future proofing" aim.
    - if you disagree with the above assessment, please do say so -- either here or on the tracking issue. It seemed like a relatively noncontroversial decision.

petrochenkov commented:
>if you disagree with the above assessment, please do say so

<!-- -->
>exhaustiveness checking should be ... precise with respect to builtin types. This is consistent with the treatment of the "builtin" constants like true, false, or 22

I disagree, literals are "variants", not "constants"
```
enum bool { true, false }
enum u8 { -128, -127, ..., 22, ..., 127 }
```
(fwiw, Haskell actually defines booleans as `data Bool = False | True`)
and there's no consistency requirement between variants and constants. Exhaustiveness checking for constants should always be conservative.

It's a very much a nit, since it currently affects only booleans (AFAIK, integers are not checked for exhaustiveness, even literals), but still.

nikomatsakis commented:
@petrochenkov 

>  Exhaustiveness checking for constants should always be conservative. (E.g. associated constants of builtin types are as real as associated constants of all other types.)

Actually, I realize now that I don't know how much we discussed the particular point of builtin types in the meeting. I think I was just projecting my personal take there. Anyway, we'll accept the RFC anyhow but I'll leave this point about builtins as unresolved for the time being. I could go either way here.

> It's a very much a nit, since it currently affects only booleans (AFAIK, integers are not checked for exhaustiveness, even literals), but still.

Yes, but it will affect constants with a type like `(true, false)`.

nikomatsakis commented:
Clean crater run finally completed. 6 regressions, as expected: https://gist.github.com/nikomatsakis/e714e4a824527e0ce5c9

nikomatsakis commented:
Update: in a recent PR (https://github.com/rust-lang/rust/pull/31277) `span` was updated to derive equality, so there should be fewer regressions than previously thought.

mrmonday commented:
Is there a nice way to deal with this, that doesn't require a loss of type safety, or far more verbose code? [libpnet](https://github.com/libpnet/libpnet) has a lot of code like:

```rust
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct EtherType(pub u16);

pub const Ipv4: EtherType = EtherType(0x0800);
// ... list of constants ...
match some_ethertype {
    Ipv4 => {},
    Other => {}
}
```

The easy way to solve it is to remove the type safety, which is what @wictory's pull request did initially - these constants have been specifically designed like this though to minimise errors (there are lots of magical numbers for lots of different things), whilst still allowing the ability to use random values if you explicitly qualify it.

eddyb commented:
@mrmonday But that example [compiles just fine](https://play.rust-lang.org/?gist=0ca5edf5489cd90f99a60a03535e4264&version=nightly) (I added more variants to make sure there wasn't an exhaustiveness-related special casing).
@bluss pointed out on IRC that if you [implement `Eq` manually](https://play.rust-lang.org/?gist=42c97763b15604907fd7516660527f08&version=nightly) you do get the errors, maybe that's what `pnet` is doing?

eddyb commented:
Turns out that [`syntex` partially expands `#[derive]`](https://gist.github.com/mrmonday/57a84ed25c60a69781543c0428154db5#file-x-rs-L104-L111) so the code in libsyntax adding `#[structural_match]` on `#[derive(..., PartialEq, ..., Eq, ...)]` doesn't trigger. I'll try to fix this in the compiler.

nikomatsakis commented:
@eddyb sigh. I can readily fix the code to have two distinct attributes, I was just being lazy.

nikomatsakis commented:
Although, triggering on `#[derive_Eq]` is a *bit* harder. Hopefully not that hard. I'm trying to remember why I chose not to do so -- I feel like there was a reason for that.
